src/BrainPass/BrainPass.sol

```diff
@@ -0,0 +1,343 @@
+    /// Constructor
+    /// -----------------------------------------------------------------------
+    
+    constructor(address IqAddr) ERC721("BAINPASS", "BEP") Owned(msg.sender) {
+        iqToken = IqAddr;
+    }
+
+    /// -----------------------------------------------------------------------
+    /// External functions
+    /// -----------------------------------------------------------------------
+
+    /// @notice Add a new Pass Type
+    /// @param pricePerDay the price per day of the new pass type 
+    /// @param tokenURI the link that stores the data of all the Nfts in the new pass
+    /// @param name the name of the new pass type to be added
+    /// @param maxTokens the total number of tokens in the pass
+    /// @param discount the amount in % to be deducted when buying the pass
+    function addPassType(
+        uint256 pricePerDay,
+        string memory tokenURI,
+        string memory name,
+        uint256 maxTokens,
+        uint256 discount
+    ) external onlyOwner {
+        if (maxTokens <= 0) revert InvalidMaxTokensForAPass();
+        uint256 passId = passIdTracker.current();
+        passTypes[passId] = PassType(
+            passId,
+            name,
+            pricePerDay,
+            tokenURI,
+            maxTokens,
+            discount,
+            0
+        );
+        passIdTracker.increment();
+        emit NewPassAdded(passId, name, maxTokens, pricePerDay);
+    }
+
+    /// @notice Mint and NFT of a particular passtype
+    /// @param passId The id of the passtype to mint
+    /// @param startTimestamp The time when the NFT subcription time starts
+    /// @param endTimestamp The time when the NFT subcription time ends
+    function mintNFT(
+        uint256 passId,
+        uint256 startTimestamp,
+        uint256 endTimestamp
+    ) external {
+        if (getUserPassDetails(msg.sender, passId).tokenId != 0)
+            revert AlreadyMintedThisPass();
+
+        PassType storage passType = passTypes[passId];
+
+        if (passType.maxTokens == 0) revert PassTypeNotFound();
+        if (passType.lastTokenIdMinted >= passType.maxTokens)
+            revert PassMaxSupplyReached();
+
+        if (!validatePassDuration(startTimestamp, endTimestamp))
+            revert DurationNotInTimeFrame();
+
+        uint256 price = calculatePrice(passId, startTimestamp, endTimestamp);
+
+        bool success = IERC20(iqToken).transferFrom(msg.sender, owner, price);
+        if (!success) revert MintingPaymentFailed();
+
+        uint256 tokenId = passType.lastTokenIdMinted + 1;
+        UserPassItem memory purchase = UserPassItem(
+            tokenId,
+            passId,
+            startTimestamp,
+            endTimestamp
+        );
+        addressToNFTPass[msg.sender][tokenId] = purchase;
+        ownerOfToken[passId][msg.sender] = purchase;
+        passType.lastTokenIdMinted = tokenId;
+
+        _safeMint(msg.sender, tokenId);
+        setBaseURI(passType.tokenURI);
```

Comment:
i think i already pointed this error. If you setBaseURI here, all nfts will have same token uri. you are changing this to all.

also move safeMint to the last step to avoid possible reentrancies